### PR TITLE
Add dedicated docs page for init-module

### LINKED ISSUE #18083: Revamp model parallel docs in PL
### 📚 Documentation

- Make an entire standalone page alone just for efficient initialization.
- Mention only the relevant part about "configure_model" in the model parallel docs.
- Add relevant info to the mixed precision docs page
- Add relevant info in the model checkpoint loading docs

That suggestion is also valid for Fabric, where we can do a better job too I think. The only difference is that we don't yet have a model parallel doc

Idea by @awaelchli as a follow-up to https://github.com/Lightning-AI/lightning/pull/18004 

cc @borda

### LINKED ISSUE #18216: Document meta device initialization for FSDP
### 📚 Documentation

Document the meta device initialization introduced in #18122 and address the open questions/limitations mentioned in the description there. 

cc @borda @carmocca @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #18083 
Fixes #18216

Adds a standalone documentation guide for `init_module` and its use cases.
Covered for both Fabric and Trainer.
Linked in pages where appropriate.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

File: docs/source-fabric/api/fabric_methods.rst [hunk 1]
```diff
@@ -157,18 +157,7 @@ This eliminates the waiting time to transfer the model parameters from the CPU t
 For strategies that handle large sharded models (FSDP, DeepSpeed), the :meth:`~lightning.fabric.fabric.Fabric.init_module` method will allocate the model parameters on the meta device first before sharding.
 This makes it possible to work with models that are larger than the memory of a single device.
 
-When loading a model from a checkpoint, for example when fine-tuning, set `empty_init=True` to avoid expensive
```
Comment (awaelchli) [inline docs/source-fabric/api/fabric_methods.rst:160]: I'm shortening this for the overview since we have a dedicated page now.

File: docs/source-fabric/advanced/model_init.rst [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    # Recommended for FSDP and DeepSpeed
+    with fabric.init_module(empty_init=True):
+        model = GPT3()  # parameters are placed on the meta-device
+
+    model = fabric.setup(model)  # parameters get sharded and initialized at once
+
+.. note::
+    Empty-init is experimental and the behavior may change in the future.
+    For FSDP, it is required that all user-defined modules that manage parameters implement a ``reset_parameters()`` method (all PyTorch built-in modules have this too).
```
Comment (carmocca) [inline docs/source-fabric/advanced/model_init.rst:75]: Mention that this is when using PyTorch 2.1+

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: docs/source-fabric/advanced/model_init.rst (hunk 2 above)
Comment (awaelchli) [reply docs/source-fabric/advanced/model_init.rst]: all done

File: docs/source-fabric/advanced/model_init.rst [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+----
+
+
+********************************************
+Model-parallel training (FSDP and DeepSpeed)
+********************************************
+
+When training sharded models with :doc:`FSDP <model_parallel/fsdp>` or DeepSpeed, using :meth:`~lightning.fabric.fabric.Fabric.init_module` is necessary in most cases because otherwise model initialization gets very slow (minutes) or (and that's more likely) you run out of CPU memory due to the size of the model.
+
+.. code-block:: python
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply docs/source-fabric/advanced/model_init.rst:70]: That's a good idea. I'll send a PR that adds this check to the method right away.

File: docs/source-fabric/advanced/model_init.rst [hunk 4]
```diff
@@ -53,6 +53,11 @@ When loading a model from a checkpoint, for example when fine-tuning, set ``empt
     model.load_state_dict(checkpoint["state_dict"])
 
 
+.. warning::
```
Comment (carmocca) [inline docs/source-fabric/advanced/model_init.rst:56]: Add this to the Trainer docs too please!
